### PR TITLE
Add parameter to power-tune only cold starts

### PR DESCRIPTION
**Pull request by mriccia**
Implement [#176](https://github.com/alexcasalboni/aws-lambda-power-tuning/issues/176) 

This is another implementation based on https://github.com/alexcasalboni/aws-lambda-power-tuning/pull/177

With this implementation the mechanism for publishing new versions is decoupled, and the Step Function workflow implements a loop to publish the versions sequentially. 
This is done to avoid hitting the Timeout on the Lambda function publishing the Versions.

### CODE REVIEW

File: test/unit/test-lambda.js
Comment (alexcasalboni) [inline test/unit/test-lambda.js:298]: This default event structure seems repeated quite a few times, it's probably worth to define as a variable and then reuse for multiple tests.

**Review by alexcasalboni — CHANGES_REQUESTED** [review]
Finally reviewed this! Thank you again for the amazing contribution 🙏 

I have quite a few comments & questions, and I still need to merge the latest master branch into this.

**Comment by alexcasalboni** [comment]
Ok, merging the latest wasn't too bad after all :)

File: lambda/utils.js
Comment (mriccia) [reply lambda/utils.js:159]: If we continue with envVars, I will investigate the need of additional permissions for kms

File: terraform/module/json_files/executor.json
Comment (mriccia) [reply terraform/module/json_files/executor.json:8]: Link to the `terraform` repo

File: lambda/initializer.js
Comment (mriccia) [reply lambda/initializer.js]: now renamed `lambdaConfigurations` as it should be clearer (as there's already an object called `initConfigurations` set in the initialisation phase)

File: lambda/utils.js
Comment (mriccia) [reply lambda/utils.js]: Ported to SDKv3 with the latest merge.
SnapStart-enabled Lambdas take longer to be published, and also there's a limit to how many versions can be published at the same time with Lambdas. 
I found through some testing that these retry options avoid/recover from failures. 

But let's discuss if we should reduce these values and/or whether we make them configurable

Comment (alexcasalboni) [reply lambda/utils.js]: @mriccia 20 retries with 500ms delay should sum up to 10 seconds - is that enough for SnapStart-enabled functions? 🤔

Comment (mriccia) [reply lambda/utils.js]: It would depend on how many versions are being published, and how long the SnapStart initialisation takes.

Comment (mriccia) [reply lambda/utils.js]: I don't have data for exactly how many, this is why I think may make sense to make these values configurable

**Comment by mriccia** [comment]
Implemented the Terraform changes in the separate repo, happy to open a PR once we have the green light on the changes here
https://github.com/mriccia/terraform-aws-lambda-power-tuning/commit/86fa3681dbe8322d229dd96c4023304fe93bc3be

File: lambda/utils.js
Comment (alexcasalboni) [inline lambda/utils.js:113]: Please add a quick comment about this. When does it happen? Is it only when restoring the power value back to its original value?

File: lambda/utils.js
Comment (alexcasalboni) [inline lambda/utils.js:178]: As mentioned in another comment, this implementation of `getLambdaPower` has become very similar to `getLambdaConfig` below. It used to make more sense because it was a specialized version that only cared about `config.MemorySize` and returned a single value. But now it returns a structure with various config parameters, so I'd consider merging `getLambdaPower` and `getLambdaConfig` back into one single implementation where we fetch all the config parameters we need.

File: lambda/utils.js
Comment (alexcasalboni) [inline lambda/utils.js:219]: Please add a quick comment/reminder of why we're doing this (for future readers/debuggers).

File: template.yml
Comment (alexcasalboni) [inline template.yml:141]: I think the publisher does NOT need the `lambda:GetFunctionConfiguration` permission now (that call is performed by the initializer).

File: lambda/executor.js
Comment (mriccia) [reply lambda/executor.js:200]: I think there's a problem with this. The cost is derived from the `durations` array, now that we're adding the `initDuration` in case of cold start the cost calculation will be inaccurate.

File: test/unit/test-lambda.js
Comment (mriccia) [reply test/unit/test-lambda.js:549]: Added some tests

File: lambda/utils.js
Comment (mriccia) [reply lambda/utils.js:52]: added some tests now

File: lambda/initializer.js
Comment (mriccia) [reply lambda/initializer.js:31]: Good point, no it's not required. Removed.

File: lambda/utils.js
Comment (mriccia) [reply lambda/utils.js:114]: yes that's correct. added a comment

File: template.yml
Comment (mriccia) [reply template.yml:140]: `lambda:GetFunctionConfiguration` is used by the `waitUntilFunctionUpdated` SDK call to check whether the Lambda function has been updated, therefore it's still required here

File: lambda/executor.js
Comment (mriccia) [reply lambda/executor.js]: I changed the way the duration and the cost are calculated. 
Please take a look when you have a moment, there are some Tests missing (planning to add early next week), but should be good to go if you're happy with this.

**Review by alexcasalboni — APPROVED** [review]

**Pull request merged**
